core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java

```diff
@@ -104,37 +170,47 @@ public GroupByNumericArrayIterator(final QueryNode node,
           + ((GroupByConfig) node.config()).getAggregator());
     }
     
-    // gotta do it here otherwise we won't know if there is any data.
-    for (final TimeSeries source : sources) {
-      if (source == null) {
-        throw new IllegalArgumentException("Null time series are not "
-            + "allowed in the sources.");
-      }
-      final Optional<TypedTimeSeriesIterator<? extends TimeSeriesDataType>> optional =
-          source.iterator(NumericArrayType.TYPE);
-      if (optional.isPresent()) {
-        final TypedTimeSeriesIterator<? extends TimeSeriesDataType> iterator = optional.get();
-        if (iterator.hasNext()) {
-          has_next = true;
-          final TimeSeriesValue<NumericArrayType> array = 
-              (TimeSeriesValue<NumericArrayType>) iterator.next();
-          // skip empties.
-          if (array.value().end() - array.value().offset() > 0) {
-            if (array.value().isInteger()) {
-              if (array.value().longArray().length > 0) {
-                aggregator.accumulate(array.value().longArray(), 
-                    array.value().offset(), array.value().end());
-              }
-            } else if (array.value().doubleArray().length > 0) {
-              aggregator.accumulate(array.value().doubleArray(),
-                  array.value().offset(), array.value().end());
+    List<Future<TimeSeriesValue<NumericArrayType>>> futures = new ArrayList<Future<TimeSeriesValue<NumericArrayType>>>(sources.size());
```

Comment:
Need a threshold to determine if we should parallelize. E.g. 1k series or greater.